src/Catalog/CatalogIndexReader.cs

```diff
@@ -86,37 +43,58 @@ public JObject GetContext()
                 pages.Add(new Tuple<DateTime, Uri>(DateTime.Parse(item["commitTimeStamp"].ToString()), new Uri(item["@id"].ToString())));
             }
 
-            return GetEntries(pages.Select(p => p.Item2));
+            return await GetEntriesAsync(pages.Select(p => p.Item2));
         }
 
-        private ConcurrentBag<CatalogIndexEntry> GetEntries(IEnumerable<Uri> pageUris)
+        private async Task<ConcurrentBag<CatalogIndexEntry>> GetEntriesAsync(IEnumerable<Uri> pageUris)
         {
-            ConcurrentBag<CatalogIndexEntry> entries = new ConcurrentBag<CatalogIndexEntry>();
+            var pageUriBag = new ConcurrentBag<Uri>(pageUris);
+            var entries = new ConcurrentBag<CatalogIndexEntry>();
+            var interner = new StringInterner();
 
-            ParallelOptions options = new ParallelOptions();
-            options.MaxDegreeOfParallelism = 8;
+            var tasks = Enumerable
+                .Range(0, MaxDegreeOfParallelism)
+                .Select(i => ProcessPageUris(pageUriBag, entries, interner))
+                .ToList();
 
-            Parallel.ForEach(pageUris.ToArray(), options, uri =>
-            {
-                var task = _httpClient.GetJObjectAsync(uri);
-                task.Wait();
+            await Task.WhenAll(tasks);
 
-                JObject json = task.Result;
+            return entries;
+        }
+
+        private async Task ProcessPageUris(ConcurrentBag<Uri> pageUriBag, ConcurrentBag<CatalogIndexEntry> entries, StringInterner interner)
+        {
+            await Task.Yield();
+            Uri pageUri;
+            while (pageUriBag.TryTake(out pageUri))
+            {
+                var json = await _httpClient.GetJObjectAsync(pageUri);
 
                 foreach (var item in json["items"])
                 {
-                    var entry = new CatalogIndexEntry(new Uri(item["@id"].ToString()),
-                            item["@type"].ToString(),
-                            item["commitId"].ToString(),
-                            DateTime.Parse(item["commitTimeStamp"].ToString()),
-                            item["nuget:id"].ToString(),
-                            NuGetVersion.Parse(item["nuget:version"].ToString()));
+                    // This string is unique.
+                    var id = item["@id"].ToString();
+
+                    // This string is discarded quickly, since DateTime.Parse does not hold a reference.
```

Comment:
It wouldn't clone the string :)

This comment is confusing though. Previously I had `item["commitTimeStamp"].ToObject<string>()`.